### DISCUSSION TITLE
WARNING (drift): When these two databases use the same QueryExecutor

### DISCUSSION THREAD

**Discussion by JerContact** [simolus3/drift#2484] [Q&A]
I just had a question about this warning to make sure i understand it!  When you mention it might create "race conditions that might corrupt the database", is that only for writes?  I just want to ask if I would be at risk of corruption if I use this technique for around 20-30 reads in very quick succession / same time?

Thanks!

**Comment by simolus3** [comment]
> is that only for writes?

The only thing that can race are schema migrations, which don't run atomically. Drift runs a schema migration like this:

1. Read saved schema version from database.
2. Run a migration if it's lower than the one declared in the database class.
3. Save the schema version to the database.

Since migrations can't safely run in transactions, these steps don't run atomically. So if you open two databases at roughly the same time, it's possible for both of them to run the migrations. This can happen even if one of the databases is only doing reads.

It's safe to open two instances of the database if you make sure to open them one at a time though, is that possible in your case?

**Reply by JerContact** [reply]
Oh great!  Yeah, all we are doing is reading from the database and any migration we would disable that ability.


Essentially each letter we type is needing to search the database, and since our database is 3 million rows it’s slowing down in the general use of drift.  We could create our own isolate manager, but seems like just having a new connection each letter would get us the results without a big refactor, at least that’s what we are thinking, and if the warning isn’t in regards to reading only, seems like a suitable solution, right?

**Reply by simolus3** [reply]
That should work since the "main" database is always open before a user is able to start typing and you start opening these short-lived databases, right? If that's the case, you can safely disable the warning.

**Reply by JerContact** [reply]
Yep, it would work exactly like that!  Ok, great, I’ll test it all out, thanks for the help!!

**Reply by JerContact** [reply]
The only problem i have now is when i close the database it's closes the main connection as well.  I assume that isn't a big deal....

And i do seem to get this

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: SqliteException(261): while executing, database is locked, database is locked (code 261)

**Reply by simolus3** [reply]
> The only problem i have now is when i close the database it's closes the main connection as well. I assume that isn't a big deal....

That actually doesn't sound great. How are you opening the main and the off-thread databases?

> [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: SqliteException(261): while executing, database is locked, database is locked (code 261)

Since one of the databases is only reading, you could try enabling [WAL mode](https://sqlite.org/wal.html) with `PRAGMA journal_mode=WAL;`. That format allows concurrent reads and writes, so since one database is read-only that could fix the issue.

**Reply by JerContact** [reply]
Right now I just have a repository class that opens a connection, and just re-instantiating that over and over....which I think is just bad, haha.

Would you suggest using "computerWithDatabase" or "serialzableConnection"?  Or do you suggest just making a whole isolate manager and doing it with managing our own isolates?  The reason it's a bit more complicated is b/c each time the user types we want the call to the database to happen immediately without waiting on other requests, so that we don't get a backup of calls to sqlite if the user types really fast.

Thanks for all the help!

Funny enough, we already have PRAGMA journal_mode=WAL; on and working

**Reply by simolus3** [reply]
> Would you suggest using "computerWithDatabase" or "serialzableConnection"?

What these methods are doing is creating a cross-isolate channel to relay database operations. So you wouldn't actually be using a second database connection in that case, you're just re-using the first one from a different isolate. Since there is only one connection in the end and drift serializes it for you, there won't be any "database is locked" errors.
However, then there's also just one thread accessing sqlite3, which if I understood you correctly lead to subpar performance for your database, right?

Are you using a setup like `NativeDatabase.createInBackground` for the repository? If not, introducing that and then trying out `computeWithDatabase` could solve the issue. If you're already using a background isolate and the performance is still not what you expect, it's unlikely that these methods will help.

**Reply by JerContact** [reply]
Yeah, right now I'm just doing what is suggested on the website:

DriftDatabaseRepository class has this:

```
DriftDatabase() : super(_openConnection());

LazyDatabase _openConnection() {
  return LazyDatabase(() async {
    final dbDirectory = await getDatabaseDirectory();
    final file = p.join(dbDirectory, 'TEST.sqlite');
    return NativeDatabase.createInBackground(
      File(file),
      setup: (database) {
        database.execute('PRAGMA journal_mode=WAL;');
        database.execute('PRAGMA foreign_keys=ON;');
        database.execute('PRAGMA recursive_triggers = ON;');
      },
    );
  });
}
```

And in the function that calls the database I'm doing this:

```
DriftDatabaseRepository repositoryTest = DriftDatabaseRepository();

// Do a bunch of work

await repositoryTest.closeConnection();
```

So I assume from what you are saying, the advanced way of making my own isolate and managing that is probably the only way forward for performance, right?
